### PR TITLE
Resolve input/output via IDuplexPipe interface once

### PR DESCRIPTION
**Pull request by benaadams**
Rather than per iteration

Highlighted calls

image

image

/cc @davidfowl @halter73 @pakrym

### CODE REVIEW

**Comment by benaadams** [comment]
```
warning CS8002: Referenced assembly 'WebDriver, Version=3.12.1.0, Culture=neutral, PublicKeyToken=null' does not have a strong name. [D:\a\1\s\src\Servers\Kestrel\test\Interop.FunctionalTests\Interop.FunctionalTests.csproj]
warning CS8002: Referenced assembly 'Pomelo.EntityFrameworkCore.MySql, Version=2.1.1.0, Culture=neutral, PublicKeyToken=null' does not have a strong name. [D:\a\1\s\src\Mvc\benchmarkapps\BasicApi\BasicApi.csproj]
warning CS8002: Referenced assembly 'Pomelo.EntityFrameworkCore.MySql, Version=2.1.1.0, Culture=neutral, PublicKeyToken=null' does not have a strong name. [D:\a\1\s\src\Mvc\benchmarkapps\BasicViews\BasicViews.csproj]
2019-01-04T02:12:47.8847130Z D:\a\1\s/src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj : warning NU1503: Skipping restore for project 'D:\a\1\s/src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj'. The project file may be invalid or missing targets required for restore. [C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.targets]
SiteExtensions\LoggingBranch\LoggingBranch.csproj :
 warning NU1503: Skipping restore for project 'D:\a\1\s/src\SiteExtensions\LoggingBranch\LoggingBranch.csproj'.
 The project file may be invalid or missing targets required for restore. [C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.targets]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\AspNetCore\AspNetCore.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\OutOfProcessRequestHandler\OutOfProcessRequestHandler.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\AspNetCore\AspNetCore.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\OutOfProcessRequestHandler\OutOfProcessRequestHandler.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\AspNetCore\AspNetCore.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\OutOfProcessRequestHandler\OutOfProcessRequestHandler.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\AspNetCore\AspNetCore.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]
warning : The referenced project 'D:\a\1\s/src\Servers\IIS\src\AspNetCoreModuleV2\OutOfProcessRequestHandler\OutOfProcessRequestHandler.vcxproj' does not exist.
 [D:\a\1\s\src\SiteExtensions\Runtime\Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj]

test\Identity.FunctionalTests\Microsoft.AspNetCore.Identity.FunctionalTests.csproj(42,5): 
 error : Unable to find precompiled view file bin\Release\netcoreapp3.0\Microsoft.AspNetCore.Identity.UI.Views.V3.dll
2019-01-04T02:12:47.8853580Z D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20181214.1\modules\vstest\module.targets(102,5):
 error : Test group UnitTests/netcoreapp3.0 failed with exit code '1'. [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20181214.1\KoreBuild.proj]
 
2019-01-04T02:12:47.8853651Z     13 Warning(s)
2019-01-04T02:12:47.8853693Z     2 Error(s)
```

**Review by pakrym — APPROVED** [review]

**Comment by benaadams** [comment]
Raised issue for the CI error, looks so close to clean CI otherwise https://github.com/aspnet/AspNetCore/issues/6389

File: src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs [hunk 1]
```diff
@@ -54,10 +54,11 @@ public Http1Connection(HttpConnectionContext context)
                 _context.TimeoutControl,
                 this);
 
+            Input = _context.Transport.Input;
             Output = _http1Output;
         }
 
-        public PipeReader Input => _context.Transport.Input;
+        public PipeReader Input { get; private set; }
```
Comment (davidfowl) [inline src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs]: nit: 

```
public PipeReader Input { get; }
```

Comment (benaadams) [reply src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs]: Done

File: src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs [hunk 2]
```diff
@@ -154,13 +154,15 @@ private async Task DoReceive()
 
         private async Task ProcessReceives()
         {
+            // Resolve `input` PipeWriter via the IDuplexPipe interface prior to loop start
+            var input = Input;
```
Comment (halter73) [inline src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs]: Nit: I would mention in the comment that this is for perf not correctness.

**Review by halter73 — APPROVED** [review]

File: src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs (hunk 2 above)
Comment (benaadams) [reply src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs]: Done

**Comment by benaadams** [comment]
Rebased to see if CI is happy yet

**Comment by benaadams** [comment]
Failed trying to sign a .js file?
```
error : Could not determine certificate name for signable file(s): [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
error : 	File: D:\a\1\s\obj\ContainerSigning\1909\tools/blazor/blazor.server.js
```
```
KoreBuild.Tasks\CodeSign.targets(51,5): error : Could not determine certificate name for signable file(s): [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error : 	File: D:\a\1\s\obj\ContainerSigning\1909\tools/blazor/blazor.server.js [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018: The "Microsoft.DotNet.SignTool.SignToolTask" task failed unexpectedly. [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018: System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary. [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018:  [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018: Server stack trace:  [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018:    at System.Collections.Generic.Dictionary`2.get_Item(TKey key) [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018:    at Microsoft.DotNet.SignTool.Configuration.GenerateListOfFiles() [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018:    at Microsoft.DotNet.SignTool.SignToolTask.ExecuteImpl() [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018:    at Microsoft.DotNet.SignTool.SignToolTask.Execute() [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018:    at System.Runtime.Remoting.Messaging.StackBuilderSink._PrivateProcessMessage(IntPtr md, Object[] args, Object server, Object[]& outArgs) [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018:    at System.Runtime.Remoting.Messaging.StackBuilderSink.SyncProcessMessage(IMessage msg) [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018:  [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018: Exception rethrown at [0]:  [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018:    at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg) [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018:    at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type) [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018:    at Microsoft.Build.Framework.ITask.Execute() [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
KoreBuild.Tasks\CodeSign.targets(51,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext() [D:\a\.dotnet\buildtools\korebuild\3.0.0-build-20190104.3\KoreBuild.proj]
```

**Comment by benaadams** [comment]
Rebased to see if CI is happy yet

**Comment by benaadams** [comment]
Close, an .IIS.FunctionalTests.AppOfflineTests failed
```
Failed   Microsoft.AspNetCore.Server.IIS.FunctionalTests.AppOfflineTests.AppOfflineDroppedWhileSiteStarting_SiteShutsDown_InProcess
Error Message:
 Assert.Equal() Failure
Expected: 0
Actual:   -1073740767
Stack Trace:
   at Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests.Helpers.AssertWorkerProcessStop(IISDeploymentResult deploymentResult, Nullable`1 timeout) in /_/src/Servers/IIS/IIS/test/Common.FunctionalTests/Utilities/Helpers.cs:line 161
   at Microsoft.AspNetCore.Server.IIS.FunctionalTests.AppOfflineTests.AppOfflineDroppedWhileSiteStarting_SiteShutsDown_InProcess() in /_/src/Servers/IIS/IIS/test/Common.FunctionalTests/AppOfflineTests.cs:line 137
--- End of stack trace from previous location where exception was thrown ---
```

**Pull request merged**